client/transaction-pool/src/revalidation.rs

```diff
@@ -0,0 +1,312 @@
+// Copyright 2018-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Pool periodic revalidation.
+
+use std::{sync::Arc, pin::Pin, collections::{HashMap, HashSet, BTreeMap}};
+
+use sc_transaction_graph::{ChainApi, Pool, ExHash, NumberFor, ValidatedTransaction, Transaction};
+use sp_runtime::traits::{Zero, SaturatedConversion};
+use sp_runtime::generic::BlockId;
+use sp_runtime::transaction_validity::TransactionValidityError;
+
+use futures::{prelude::*, channel::mpsc, stream::unfold};
+use std::time::Duration;
+use futures_timer::Delay;
+
+#[cfg(not(test))]
+const BACKGROUND_REVALIDATION_INTERVAL: Duration = Duration::from_millis(200);
+#[cfg(test)]
+const BACKGROUND_REVALIDATION_INTERVAL: Duration = Duration::from_millis(5);
+
+const BACKGROUND_REVALIDATION_BATCH_SIZE: usize = 20;
+
+/// Payload from queue to worker.
+struct WorkerPayload<Api: ChainApi> {
+	at: NumberFor<Api>,
+	transactions: Vec<ExHash<Api>>,
+}
+
+/// Async revalidation worker.
+///
+/// Implements future and can be spawned in place or in background.
+struct RevalidationWorker<Api: ChainApi> {
+	api: Arc<Api>,
+	pool: Arc<Pool<Api>>,
+	best_block: NumberFor<Api>,
+	block_ordered: BTreeMap<NumberFor<Api>, HashSet<ExHash<Api>>>,
+	members: HashMap<ExHash<Api>, NumberFor<Api>>,
+}
+
+impl<Api: ChainApi> Unpin for RevalidationWorker<Api> {}
+
+fn interval(duration: Duration) -> impl Stream<Item=()> + Unpin {
+	unfold((), move |_| {
+		Delay::new(duration).map(|_| Some(((), ())))
+	}).map(drop)
+}
+
+/// Actual async revalidation batch.
```

Comment:
Please elaborate, please at least attempt to roughly define the process of what `async revalidation batch` is supposed to be.